Optimise creation of anonymous functions

michalmuskala opened the pull request:
This intorduces a similar optimisation for normal funs
to what was introduced for external funs in #1725.
It is possible to allocate the fun as a literal, if it does not capture
the environemt (i.e. it does not close over any variables).
Unfortunately it's not possible to do this in the compiler due to
problems with representation of such functions in the `.beam` files.
Fortunately, we can do this in the loader.

Simple evaluation shows that functions that don't capture the
enviornment consistute over 60% of all funs in the source code of
Erlang/OTP itself.

The only downside is that we lose a meningful value in the `pid` field
of the fun. The goal of this field, beyond debugging, was to be
able to identify the original node of a function. To be able to still do
this, the functions that are created in the loader are assigned the init
pid as the creator.

gootik commented:
I wonder how this would change benchmarks.

bjorng commented:
Looks simple and safe enough that we probably can include it in RC2 for OTP 21.

Please update the documentation for `{pid, Pid}` in the `erlang:fun_info/1` documentation. While at it, you can fix the spelling errors in the commit message ("intorduces" and "enviornment").

We will review this pull request closer on Monday next week. (Today, Thursday, is a public holiday in Sweden, and tomorrow is also a holiday.)

bjorng commented:
@michalmuskala I have pushed a fixup commit with additional tests of literal funs. Please squash it into your commit.

bjorng commented:
Otherwise looks good. We will add the PR to our daily builds when you have fixed the issues mentioned above.

michalmuskala commented:
Updated and squashed. I think this should fix the `crashdump_viewer_SUITE` error as well, but I wasn't able to run the suite at all to verify it works correctly (it fails in `init_per_SUITE` for me).

michalmuskala commented:
Updated.

bjorng commented:
Added to our daily builds.

garazdawi commented:
Something in this PR causes termination of nodes to segfault. So far we've gotten about 300 cores last night with this or very similar backtraces:

```
#0  0x081683c0 in enc_pid (acmp=0x0, pid=0, ep=0xb4efdb09 "", dflags=397188)
    at beam/external.c:2254
#1  0x0816902b in enc_term_int (ctx=0x0, acmp=0x0, obj=83, ep=0xb4efdb09 "", 
    dflags=397188, off_heap=0x0, reds=0x0, res=0xb4efdab4)
    at beam/external.c:2951
#2  0x0816a940 in erts_encode_ext (term=3058436558, ext=0xb4efdedc)
    at beam/external.c:2444
#3  0x081cb7ee in dump_externally (to=0x8282d50 <erts_write_fp>, 
    to_arg=0x833ca48, term=3035618057) at beam/erl_process_dump.c:715
#4  0x081cd0eb in erts_deep_process_dump (to=0x8282d50 <erts_write_fp>, 
    to_arg=0x833ca48) at beam/erl_process_dump.c:967
#5  0x081b3160 in erl_crash_dump_v (file=0x0, line=0, fmt=0x82befe7 "%s\n", 
    args=0xb4eff0c8 "\200�0\b��������M�\220��") at beam/break.c:953
#6  0x080e387c in erts_exit_vv (n=-3, flush_async=<value optimized out>, 
    fmt=0x82befe7 "%s\n", args1=0xb4eff0c8 "\200�0\b��������M�\220��", 
    args2=0xb4eff0c8 "\200�0\b��������M�\220��") at beam/erl_init.c:2406
#7  0x080e38bd in erts_exit (n=-3, fmt=0x82befe7 "%s\n")
    at beam/erl_init.c:2433
#8  0x081243da in halt_2 (A__p=0xb6740088, BIF__ARGS=0xb6d400c0, 
    A__I=0xb5ba58c0) at beam/bif.c:3784
#9  0x080818b9 in process_main (x_reg_array=0xb6d400c0, f_reg_array=0xb6d41100)
    at i686-pc-linux-gnu/opt/smp/beam_hot.h:310
#10 0x0808cd8d in sched_thread_func (vesdp=0xb5a022c0)
    at beam/erl_process.c:8349
#11 0x0827f2ab in thr_wrapper (vtwd=0xbf826488) at pthread/ethread.c:118
#12 0xb77721b5 in start_thread () from /lib/libpthread.so.0
#13 0xb76d762e in clone () from /lib/libc.so.6
```

michalmuskala commented:
Looking at the stacktrace the immediate problem seems to be the value of `pid` being 0, which seems surprising because the calling code switches on `tag_val_def(pid)`. I'll try to investigate why that might be happening.

michalmuskala commented:
I think the problem is that some modules are loaded with this optimisation before the `erts_init_process_id` value is set up, this means they get assigned not the init process, but value 0.

I'm not sure what's the best course of action here. I could see two solutions:
* if `erts_init_process_id` is 0, set the `creator` value to something else (maybe atom `undefined`).
* if `erts_init_process_id` is 0, skip the optimisation and emit regular `make_fun` instruction.

Do you think any of the solutions is preferred or maybe there's another way to solve this?

bjorng commented:
@michalmuskala Skipping the optimization seems to be the cleaner solution.

michalmuskala commented:
I pushed a commit that initialises the value of `erts_init_process_id` to `ERTS_INVALID_PID` instead of leaving it uninitialised. That's another possible approach, but I'm not sure what other repercussions this could have in the system.

jhogberg commented:
Looks good! I'll merge it after you've squashed the commits.

fenollp commented:
What about also adding a warning that says something along the lines of
> This function does not close over the environment and should be made into an un-exported named function.

Or maybe I just *really* want a linter :)

michalmuskala commented:
@jhogberg squashed

@fenollp I don't think such a warning would be helpful - we use such functions extremely often, for example: `lists:map(fun(X) -> X * 2 end, List)` - would it really be beneficial to make this a named function in all such cases?
